# Question 3

1. In this part, you are required to implement Apriori and FPgrowth algorithms on the data provided as "data_apriori_fpgrowth/data.csv". The dataset is of online directory of certified businesses.

2. We have loaded the dataset for you.

3. You are free to use any library or implement your own algorithms.


**Use minimum support : 0.1**

In [1]:
import pandas as pd
import numpy as np

from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
import time
from mlxtend.frequent_patterns import fpgrowth

In [2]:
data = np.load("data_apriori_fpgrowth/data.npy", allow_pickle = True)

# To Do
1. Run Apriori Algorithm on the dataset using minimum support 0.1.
 1. Note the run time of the algorithm to find frequent item sets.
 1. Print all the frequent item sets found by apriori.
 1. Print all frequent item sets of length $2$ or higher with support of $0.2$ or higher. 
1. Run FPGrowth Algorithm on the dataset using minimum support 0.1.
 1. Note the run time of the FPGrowth algorithm to find frequent item sets.
 1. Print all frequent item sets found by FPGrowth.
 1. Print all frequent item sets of length $2$ or higher with support of $0.2$ or higher.
1. Answer the following question:
 1. Both algorithms find and return exactly the same frequent itemsets (given same parameters of support and itemset length). What difference do you note in both of the algorithms?
 1. There were more than $1000$ transactions. Why there are too few itemsets returned by the algorithms?
 

# Apriori [10 marks]

In [3]:
# Preprocess data
te = TransactionEncoder()
te_ary = te.fit(data).transform(data)
df = pd.DataFrame(te_ary, columns=te.columns_)

In [4]:
# Apriori
# All Frequent Itemsets
start = time.perf_counter()
apriori(df, min_support=0.1, use_colnames=True)

support                       itemsets
0   0.202113                        (ASIAN)
1   0.300704                        (BLACK)
2   0.152113                     (Brooklyn)
3   0.164085                     (HISPANIC)
4   0.671127                          (MBE)
5   0.300000                 (NON-MINORITY)
6   0.295070                     (New York)
7   0.477465                          (WBE)
8   0.200000                   (ASIAN, MBE)
9   0.300704                   (BLACK, MBE)
10  0.112676                (Brooklyn, MBE)
11  0.164085                (HISPANIC, MBE)
12  0.170423                (New York, MBE)
13  0.169014                     (WBE, MBE)
14  0.118310       (NON-MINORITY, New York)
15  0.300000            (NON-MINORITY, WBE)
16  0.175352                (WBE, New York)
17  0.118310  (NON-MINORITY, New York, WBE)

In [5]:
# Frequent itemsets of length 2 with support of 0.2
print("Item sets with min. support 0.2 or greater and length 2 or greater")
frequent_itemsets = apriori(df, min_support=0.2, use_colnames=True)
frequent_itemsets['length'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
frequent_itemsets[(frequent_itemsets['length'] >= 2)]

Item sets with min. support 0.2 or greater and length 2 or greater


support             itemsets  length
6  0.200000         (ASIAN, MBE)       2
7  0.300704         (BLACK, MBE)       2
8  0.300000  (NON-MINORITY, WBE)       2

In [6]:
# Time taken by apriori
print("Time taken by apriori: {0:.2f} seconds".format(time.perf_counter() - start))

Time taken by apriori: 0.19 seconds


# FPGrowth [10 marks]

In [7]:
# FPGrowth
# All Frequent Itemsets
start = time.perf_counter()
fpgrowth(df, min_support=0.1, use_colnames=True)

support                       itemsets
0   0.671127                          (MBE)
1   0.477465                          (WBE)
2   0.300704                        (BLACK)
3   0.152113                     (Brooklyn)
4   0.202113                        (ASIAN)
5   0.295070                     (New York)
6   0.164085                     (HISPANIC)
7   0.300000                 (NON-MINORITY)
8   0.169014                     (WBE, MBE)
9   0.300704                   (BLACK, MBE)
10  0.112676                (Brooklyn, MBE)
11  0.200000                   (ASIAN, MBE)
12  0.175352                (WBE, New York)
13  0.170423                (New York, MBE)
14  0.118310       (NON-MINORITY, New York)
15  0.118310  (NON-MINORITY, New York, WBE)
16  0.164085                (HISPANIC, MBE)
17  0.300000            (NON-MINORITY, WBE)

In [8]:
# Frequent itemsets of length 2 with support of 0.2
print("Item sets with min. support 0.2 or greater and length 2 or greater")
frequent_itemsets = fpgrowth(df, min_support=0.2, use_colnames=True)
frequent_itemsets['length'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
frequent_itemsets[(frequent_itemsets['length'] >= 2)]

Item sets with min. support 0.2 or greater and length 2 or greater


support             itemsets  length
6  0.300704         (BLACK, MBE)       2
7  0.200000         (ASIAN, MBE)       2
8  0.300000  (NON-MINORITY, WBE)       2

In [9]:
# Time taken by FPgrowth
print("Time taken by FPgrowth: {0:.2f} seconds".format(time.perf_counter() - start))

Time taken by FPgrowth: 0.19 seconds


# Answer to the Questions [5 marks]

3A. In this case, it can be seen that Apriori takes 0.15 seconds to run, whereas FPgrowth takes 0.18 seconds to run. Apriori performs multiple scans of the database, whereas FPgrowth scans the database only twice. This is why FPgrowth is more memory and space efficient. FPgrowth is a tree-based algorithm and uses depth-first search. On the other hand, Apriori is an array-based algorithm and uses breadth-first search.

3B. Only a few itemsets are returned because the number of original items (1000) is quite low by data mining standards. Moreover, the minimum support is 0.2 (in the second case), which means that only items that appear more than 200 (1000 * 0.2) would be treated as frequent items. The minimum support is relatively high for the total number of items.